helix-core/src/main/java/org/apache/helix/messaging/handling/HelixTaskExecutor.java

```diff
@@ -714,7 +724,7 @@ private void syncSessionToController(HelixManager manager) {
 
   @Override
   @PreFetch(enabled = false)
-  public void onMessage(String instanceName, List<Message> messages,
+  public synchronized void onMessage(String instanceName, List<Message> messages,
```

Comment:
@jiajunwang 14 hours ago Contributor
Why? The periodic refresh does not need to be done synchronized, I think.
If your timer action just adds the refresh event to the queue, there is no need to modify this.

Reply:
Previously only CallbacHandler (for message) would execute this onMessage() and it would be on ZkClient's event thread. 
Now you have thread to execute this method. Make it would synchronized would be safe.

Reply:
If we make sure periodic refresh timer is reset whenever there's been recent events, this sort of synchronization won't be necessary by design. See my more detailed comment below.

Reply:
Per our conversation, moved this change from HelixTaskExecutor to CallbackHandler. Instead of call onMessage directly, for refresh, we enqueue a specific event to the queue to trigger onMessage.